esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+		w.run()
+		bi.workers = append(bi.workers, &w)
+	}
+	bi.wg.Add(bi.config.NumWorkers)
+
+	bi.ticker = time.NewTicker(bi.config.FlushInterval)
+	go func() error {
+		for {
+			select {
+			case <-bi.ticker.C:
+				if bi.config.DebugLogger != nil {
+					bi.config.DebugLogger.Printf("[indexer] Auto-flushing workers after %s\n", bi.config.FlushInterval)
+				}
+				for _, w := range bi.workers {
+					w.mu.Lock()
+					if w.buf.Len() > 0 {
+						if err := w.flush(); err != nil {
+							// TODO(karmi): Wrap error
+							w.mu.Unlock()
+							return fmt.Errorf("%s", err)
+						}
+					}
+					w.mu.Unlock()
+				}
+			}
+		}
+	}()
+}
+
+// worker represents an indexer worker.
+//
+type worker struct {
+	id    int
+	ch    <-chan BulkIndexerItem
+	mu    sync.Mutex
+	bi    *bulkIndexer
+	buf   *bytes.Buffer
+	aux   []byte
+	items []BulkIndexerItem
+
+	isFlushing bool
+}
+
+// run launches the worker in a goroutine.
+//
+func (w *worker) run() {
+	go func() {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Started\n", w.id)
+		}
+		defer w.bi.wg.Done()
+
+		for item := range w.ch {
+			w.mu.Lock()
+
+			if w.bi.config.DebugLogger != nil {
+				w.bi.config.DebugLogger.Printf("[worker-%03d] Received item [%s:%s]\n", w.id, item.Action, item.DocumentID)
+			}
+
+			if err := w.writeMeta(item); err != nil {
+				if item.OnFailure != nil {
+					item.OnFailure(item, BulkIndexerResponseItem{}, err)
+				}
+				atomic.AddUint64(&w.bi.stats.numFailed, 1)
+				w.mu.Unlock()
+				continue
+			}
+
+			if err := w.writeBody(item); err != nil {
+				if item.OnFailure != nil {
+					item.OnFailure(item, BulkIndexerResponseItem{}, err)
+				}
+				atomic.AddUint64(&w.bi.stats.numFailed, 1)
+				w.mu.Unlock()
+				continue
+			}
+
+			w.items = append(w.items, item)
+			if w.buf.Len() >= w.bi.config.FlushBytes {
+				w.flush()
+			}
+			w.mu.Unlock()
+		}
+	}()
+}
+
+// writeMeta formats and writes the item metadata to the buffer; it must be called under a lock.
+//
+func (w *worker) writeMeta(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	w.buf.WriteRune('{')
+	w.aux = strconv.AppendQuote(w.aux, item.Action)
+	w.buf.Write(w.aux)
+	w.aux = w.aux[:0]
+	w.buf.WriteRune(':')
+	w.buf.WriteRune('{')
+	if item.DocumentID != "" {
+		w.buf.WriteString(`"_id":`)
+		w.aux = strconv.AppendQuote(w.aux, item.DocumentID)
+		w.buf.Write(w.aux)
+		w.aux = w.aux[:0]
+	}
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('\n')
+	return nil
+}
+
+// writeBody writes the item body to the buffer; it must be called under a lock.
+//
+func (w *worker) writeBody(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	if item.Body != nil {
+		if _, err := w.buf.ReadFrom(item.Body); err != nil {
+			return err
+		}
+		w.buf.WriteRune('\n')
+	}
+	return nil
+}
+
+// flush writes out the worker buffer; it must be called under a lock.
+//
+func (w *worker) flush() error {
+	if w.isFlushing {
```

Comment:
I might be overlooking something, but when the callers need to safeguard that `flush` is called under a lock, then how could it happen that it is called while flushing? And if it can happen, is it then safe that there is no explicit lock around reading and writing `w.isFlushing`? 
As far as I understand, the `isFlusing` can be completely dismissed.

Reply:
Yeah, I'll need to look into it again — previously, there was no mutex synchronization, before I added the periodic flush interval, so this might only a residue of the previous implementation. (Note the flush is called from `Close()`.)

Reply:
Thanks, this is indeed not needed, removed in a129cd4!